bindings/cs/Protean-cs/Protean.Net/BinaryWriter.cs

```diff
@@ -65,357 +65,561 @@ public static byte[] ToBytes(Variant value, BinaryMode mode)
+        {
+            using (MemoryStream ms = new MemoryStream(buffer, bufferIndex, buffer.Length - bufferIndex))
+            {
+                using (BinaryWriter writer = new BinaryWriter(ms, mode))
+                {
+                    writer.Write(value);
+                }
+
+                return (int)ms.Position;
+            }
+        }
+
+        public static int GetByteCount(Variant variant)
+        {
+            return 12 + GetByteCountVariant(variant);
+        }
+
+        private void WriteHeader()
+        {
+            // Create and write the Protean header
+            m_stream.Write(BitConverter.GetBytes(BinaryConstants.PROTEAN_MAGIC), 0, 4);
+            m_stream.Write(BitConverter.GetBytes(BinaryConstants.VERSION_MAJOR << 16 | BinaryConstants.VERSION_MINOR), 0, 4);
+            m_stream.Write(BitConverter.GetBytes((int)m_mode), 0, 4);
+        }
+
+        // This doesn't seem to support Unicode properly
+        private void Write(string arg)
+        {
+            int length = arg.Length;
+            byte[] bytes = new byte[length];
+
+            for (int i = 0; i < length; ++i)
+            {
+                bytes[i] = (byte)arg[i];
+            }
+
+            Write(length);
+            Write(bytes, true);
+        }
+
+        private static int GetByteCountString(string arg)
+        {
+            return sizeof(Int32) + GetByteCountBytes(arg.Length, true);
+        }
+
+        private void Write(TimeSpan arg)
+        {
+            Write((Int64)arg.TotalMilliseconds);
+        }
+
+        private static readonly long MaxDateTimeMillis = (Variant.MaxDateTime.Ticks - Variant.MinDateTime.Ticks) / 10000;
+
+        private void Write(DateTime arg)
+        {
+            if (arg == DateTime.MaxValue)
+            {
+                Write(MaxDateTimeMillis);
+            }
+            else
+            {
+                Write((arg.Ticks - Variant.MinDateTime.Ticks) / 10000);
+            }
+        }
+
+        private void Write(byte arg)
+        {
+            m_filter.WriteByte(arg);
+        }
+
+        private void Write(Int32 arg)
+        {
+            Write(BitConverter.GetBytes(arg));
+        }
+
+        private void Write(UInt32 arg)
+        {
+            Write(BitConverter.GetBytes(arg));
+        }
+
+        private void Write(Int64 arg)
+        {
+            Write(BitConverter.GetBytes(arg));
+        }
+
+        private void Write(UInt64 arg)
+        {
+            Write(BitConverter.GetBytes(arg));
+        }
+
+        private void Write(float arg)
+        {
+            Write(BitConverter.GetBytes(arg));
+        }
+
+        private void Write(double arg)
+        {
+            Write(BitConverter.DoubleToInt64Bits(arg));
+        }
+
+        private void Write(byte[] arg)
+        {
+            m_filter.Write(arg, 0, arg.Length);
+        }
+
+        private void Write(bool arg)
```

Comment:
While you're here, could you add some more brackets in the call to `Write()` below, just to be sure? :)